### PR TITLE
Allow specification of additional module search paths via `g:deoplete#sources#jedi#site_packages`

### PR DESCRIPTION
**Pull request by coddingtonbear**
In some situations -- in my case, when running nvim within a GUI application -- a virtualenv can't be used for providing additional package paths for completions.  What this change does is allows one to set a specific configuration variable to a colon-delimited list of paths to add to Jedi's `PYTHONPATH`.  For example:

```
let g:deoplete#sources#jedi#site_packages='/some/path/to/a/special/site-packages/directory/'
```

Packages within the above directory would then become available to Jedi for completions.

### CODE REVIEW

**Comment by zchee** [comment]
@coddingtonbear Thanks for pull request!
I will check and try it later.

/cc @tweekmonster If you have time, Could you double check it?

**Comment by tweekmonster** [comment]
@zchee This PR does work, but I'm not sure if this is the best approach.

Setting `$VIRTUALENV` in `init.vim` or an `'exrc'` script will be the best way to go about this.  But, it has to be done before rplugin host starts and before deoplete loads the sources.  The reason is that `python` might be a different version from the virtualenv.  Any compiled modules in the virtualenv would cause Jedi to segfault since they were built with a different version.

@coddingtonbear Could you test if `let $VIRTUALENV = '/path/to/venv'` in `init.vim` works in the GUI nvim?

We might have to come up with a way for the virtualenv to change at runtime.  Like have a command that allows the user to restart the server with new env variables.

**Comment by coddingtonbear** [comment]
Although I could probably sort out a workaround, in my specific use case for this, there _is_ no virtualenv; I was working with packages  that are stored in my user-level package directory (the directory installed into when you get when you run `pip install --user <package>`).

**Comment by tweekmonster** [comment]
In that case try setting [PYTHONUSERBASE](https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONUSERBASE) because that would have the same problem in regard to compiled modules.

**Comment by coddingtonbear** [comment]
Neat; that does appear to work.

**Comment by coddingtonbear** [comment]
Thanks for the tip; @tweekmonster -- I think setting`PYTHONUSERBASE` in `init.vim` solves my problem completely (incidentally: I had no idea you could even _set_ environment variables in there and appreciate the useful functionality).

**Comment by coddingtonbear** [comment]
Actually, I think I've spoken too soon; when setting `PYTHONUSERBASE` to the proper env path, the Python 3 host loses its ability to start; this is maybe related to my `PYTHONUSERBASE` being a Python 2.7 environment.  Ideas?

**Comment by tweekmonster** [comment]
@coddingtonbear Are you setting `PYTHONUSERBASE` to the virtualenv or to the user site directory (e.g. `~/.local/`).  It should match `USER_BASE` in the output of `python -m site`

**Comment by coddingtonbear** [comment]
For me, `USER_BASE` in python is:

```
 $ python
Python 2.7.11 (default, Jan 22 2016, 08:28:37)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import site
>>> site.USER_BASE
'/Users/adamcoddington/Library/Python/2.7'
```

In my `init.vim`, I've added the following line:

```
let $PYTHONUSERBASE='/Users/adamcoddington/Library/Python/2.7'
```

And when doing so, the error message I get is:

```
function provider#python3#Call[9]..remote#host#Require[13]..provider#pythonx#Require, line 15
Vim(if):Channel was closed by the client
Failed to load python3 host. You can try to see what happened by starting Neovim with the environment variable $NVIM_PYTHON_LOG_FILE set to a file and opening the generated log file. Also, the host stderr will be available in Neovim log, so it may contain useful information. See also ~/.nvimlog.
```

**Comment by coddingtonbear** [comment]
If I run with `NVIM_LOG_FILE` set -- this is the content of the log file:

```
2016-08-16 10:55:52,824 [INFO @ script_host.py:setup:52] 89156 - install import hook/path
2016-08-16 10:55:52,824 [INFO @ script_host.py:setup:57] 89156 - redirect sys.stdout and sys.stderr
```

**Comment by zchee** [comment]
@tweekmonster @coddingtonbear Ah, thanks digging.
I will reading this threads. wait a moment. sorry

**Comment by tweekmonster** [comment]
@coddingtonbear I'm going to get a GUI client and see what I can find out.  Are you using the built-in OS X python interpreter?

**Comment by coddingtonbear** [comment]
I'm afraid I'm not -- in my shell I'm using 2.7.11 (from homebrew), but both return the same `USER_BASE` (as you'd expect given that they're on the same minor).

Re: GUI, I'm using [neovim-dot-app](https://github.com/rogual/neovim-dot-app) and in my `init.vim`, I've set the following interpreters:

```
let g:python_host_prog = '/usr/local/bin/python'
let g:python3_host_prog = '/usr/local/bin/python3'
```

**Comment by tweekmonster** [comment]
@coddingtonbear Alright, I took a look and it looks like the problem is that the server is using just `python` and relying on the search paths to find the appropriate interpreter.  `$PATH` with `/usr/local/bin` isn't seen by the server process, so it's using `/usr/bin/python` instead.

I bet if you run `!python3 -m site` its output will be different than what you see from running the same thing in the terminal.  And `:echo $PATH` will probably be different than `echo $PATH` in the terminal.

You can try `let $PATH = '/usr/local/bin:'.$PATH` and see if it works.  I'm still seeing if there's a better way to go about this.

**Comment by coddingtonbear** [comment]
Yeah -- running `!python3 -m site` _does_ output a different path than that (`!python -m site`, though, doesn't -- as you'd expect, of course), but the libraries I'm hoping to use completions for are for a Python 2.7 project.

**Comment by tweekmonster** [comment]
I meant in Neovim vs in the standard terminal.

**Comment by coddingtonbear** [comment]
Oh, I knew what you meant, @tweekmonster

**Comment by coddingtonbear** [comment]
I think maybe part of why this might be confusing is that although the jedi version we're running has to match the version of python executing it, it can provide completions for python code of any version.  Maybe there's another way to convince jedi into becoming aware of specific libraries, but I'm not sure what one might be aside from adding something to the `PYTHONPATH`.

**Comment by tweekmonster** [comment]
Sorry I thought you were comparing the output of `python3` vs `python`.  The point I was making is that `$PATH` is most likely not what you're expecting it to be when the GUI launches.

If you run `!which python` in Neovim, it'll probably show `/usr/bin/python`.  If you're using the homebrew version of Python, you'd want it to show `/usr/local/bin/python`.

> Maybe there's another way to convince jedi into becoming aware of specific libraries, but I'm not sure what one might be aside from adding something to the PYTHONPATH.

Setting `PYTHONPATH`, `VIRTUALENV`, or injecting into its processes' `sys.paths` are how you would have to do it, and that's basically what the server does.  The only flaw in it is that the server expects `which python` to be what you're using in the shell.

**Comment by coddingtonbear** [comment]
I'm afraid not, @tweekmonster --

```
adamcoddington@Adam-Coddington: ~/Documents/Projects/webenv master!
 $ which python                                                                                                [11:03:58]
/usr/local/bin/python

adamcoddington@Adam-Coddington: ~/Documents/Projects/webenv master!
 $ which python3                                                                                               [13:41:53]
/usr/local/bin/python3
```

In neovim-dot-app:

```
!which python
/usr/local/bin/python
```

```
!which python3
/usr/local/bin/python3
```

**Comment by coddingtonbear** [comment]
To be clear: I think a much more likely explanation is that we're running jedi using the python3 interpreter, but that given that `PYTHONUSERBASE` is a Python 2.7 base directory, the interpreter can't start.

**Comment by tweekmonster** [comment]
> I think a much more likely explanation is that we're running jedi using the python3 interpreter

That would only be the case if `python` that's being launched *is* python 3.  In any case, enable logging for deoplete and this plugin:

```
call deoplete#custom#set('django', 'debug_enabled', 1)
call deoplete#enable_logging('DEBUG', '/tmp/deoplete.log')
```

That'll write deoplete's logs to `/tmp/deoplete.log` and the Jedi server's logs at `/tmp/jedi-server.log`.  When it starts, it'll print `sys.paths` which should help figure out where it's finding completions.

**Comment by coddingtonbear** [comment]
What if some _other_ plugins I'm using might demand the Python 3 interpreter; given that the `PYTHONUSERBASE` environment variable will apply to all spawned processes (either python2 or python3), is that an acceptable outcome?

**Comment by coddingtonbear** [comment]
Your logs --

* `/tmp/jedi-server.log`: https://gist.github.com/coddingtonbear/63c9f14c17f68df4c8fb5943720fba2c
* `/tmp/deoplete.log`: https://gist.github.com/coddingtonbear/a24b0a642200b9452701bd5a72a32a1d

To clarify my above comment; the following things are true to the best of my knowledge:

* Neovim can be built with both python3 and python2 bindings simultaneously.
* Environment variables set in `init.vim` apply to all subprocesses.
* `PYTHONUSERBASE` is specific to a given python version.
* When launching neovim while `PYTHONUSERBASE` is set to point at a python2 environment, Jedi works* (because its running in the python2 interpreter) but the python3 interpreter cannot 
start because *its* `PYTHONUSERBASE` is set to point at a python2 environment, too.

Regarding the last one, that's why I originally thought your solution worked; it was only later that I discovered that the python3 interpreter couldn't start.  And, at that point, I convinced myself that the (working) completions I saw weren't from Jedi but were from one of the other completion engines.

So, TLDR: I think the problem is that setting the `PYTHONUSERBASE` environment variable will disable any python interpreters of versions other than the version for which that `PYTHONUSERBASE` was intended, and neovim can be built with bindings for both python2 and python3 simultaneously.

**Comment by tweekmonster** [comment]
> When launching neovim while PYTHONUSERBASE is set to point at a python2 environment, Jedi works* (because its running in the python2 interpreter) but the python3 interpreter cannot start because its PYTHONUSERBASE is set to point at a python2 environment, too.

Yeah sorry, I understand that now that I took a third look at the outputs of `python -m site` on my Mac.  On my linux box that I use for dev, `USER_BASE` is actually a *base* directory that doesn't include the specific Python versions (e.g. `~/.local`).  It means that I can set `PYTHONUSERBASE` to an arbitrary path for any version.  So, I had it in my head that it was a simple case.  

> Environment variables set in init.vim apply to all subprocesses.

This is true in regard to the RPC hosts being started.  Afterwards, it stops mattering because the environment can't be changed for an already-running subprocess.


One more thing you can try if you want to be absolutely sure the Python 2 interpreter is used:

```vim
let deoplete#sources#jedi#python_path = '/usr/local/bin/python2'
```

Make sure `PYTHONUSERBASE` isn't set.  If you start typing `xra` in a Python file, you should see a completion for `xrange` which isn't available in Python 3.  If you don't see this completion and the jedi server is running, I can't even begin to figure out what's going on there.

If you have the time for it, try out [pyenv](https://github.com/yyuu/pyenv).  Maybe you'll get more predictable results.  I always felt that homebrew Python is a giant PITA.

**Comment by coddingtonbear** [comment]
So, I'm not sure I follow here -- what I'm saying is that Jedi is fine if I set `PYTHONUSERBASE`, but nothing that relies on the python3 interpreter will work.  I'm not clear on how telling Jedi which python interpreter to use will work around either of those problems; care to clarify?

**Comment by tweekmonster** [comment]
I guess I'm the one that's unclear on what's proposed, then.  `PYTHONUSERBASE` shouldn't need to be set if you didn't set it when you installed your packages via `--user`.  It should just work when using the interpreter that was used by `pip`.

This will give you the paths and env that the server subprocess will see when it starts:

```vim
:python3 import os,subprocess;env=os.environ.copy();env['PYTHONPATH']='/whatever';print(subprocess.check_output("python -c \"import os,sys,sysconfig; out=['exe: '+sys.executable,''];out.extend(['Build Paths:']+sysconfig.get_config_vars('base', 'userbase', 'platbase', 'prefix', 'exec_prefix'));out.extend(['sys.path:']+sys.path);out.extend(['','Env:','']);out.extend(['%s=%s' % (x, y) for x, y in os.environ.items()]);print('\\n'.join(out))\"", shell=True, env=env).decode('utf8'))
```

It'll display the executable, the paths that would influence module search, and the environment.  Maybe there'll be a clue as to why your user packages aren't seen.  To verify that the executable actually has the user packages, run: `/path/to/python -m pip freeze --user`.

If you want to include an arbitrary path that has nothing to do with Python's default search paths, `PYTHONPATH` needs to be used, but deoplete-jedi currently overwrites it.  This PR would be partially valid (and where my understanding of the problem got split).  It just needs to be made so that an existing `PYTHONPATH` is appended instead of overwritten.

That being said, I'm currently working on refactoring the client/server so it can be used in other deoplete sources, and it doesn't completely overwrite `PYTHONPATH`.

**Comment by coddingtonbear** [comment]
I'm afraid I've already spent far too much time trying to work through this conversation; I'm just going to continue using my fork and let you guys figure out what you want to do for solving this problem when it occurs for the next person.

**Comment by tweekmonster** [comment]
Sorry we couldn't sort this out.  If you happen to figure out the exact reason why Python can only find your user packages by setting an env variable, please let us know.
